apps/src/blockly/addons/cdoUtils.js

```diff
@@ -356,3 +356,102 @@ export function partitionBlocksByType(
 
   return [...prioritizedBlocks, ...remainingBlocks];
 }
+
+/**
+ * Retrieves the toolbox blocks for a custom category from the level config.
+ * @param {string} customCategory The name of the custom category to retrieve blocks from. (Ex. 'VARIABLE', 'Behavior')
+ * @returns {Document} A new XML document containing the filtered blocks.
+ */
+export function getLevelToolboxBlocks(customCategory) {
+  const parser = new DOMParser();
+  // TODO: Update this to support JSON once https://codedotorg.atlassian.net/browse/CT-8 is merged
+  const xmlDoc = parser.parseFromString(Blockly.toolboxBlocks, 'text/xml');
+
+  // Find the category based on the custom attribute
+  const categories = xmlDoc.getElementsByTagName('category');
+  let foundCategory = null;
+
+  for (const category of categories) {
+    if (category.getAttribute('custom') === customCategory) {
+      foundCategory = category;
+      break;
+    }
+  }
+
+  if (foundCategory) {
+    // Create a new XML document and append the child nodes of the category to it
+    const newXmlDocument = parser.parseFromString(
+      '<xml></xml>',
+      'application/xml'
+    );
+    for (const childNode of foundCategory.childNodes) {
+      // Clone the child node and append it to the new XML document
+      newXmlDocument.documentElement.appendChild(childNode.cloneNode(true));
+    }
+
+    return newXmlDocument;
+  } else {
+    return undefined;
+  }
+}
+
+/**
+ * Simplifies the state of blocks for a flyout by removing properties like x/y and id.
+ * Also replaces variable IDs with variable names derived from the serialied variable map.
+ * @param {object} serialization The full workspace serialization, including variables and blocks.
+ * @returns {Array<object>} An array of simplified block objects, suitable for a flyout.
+ */
+export function getSimplifiedStateForFlyout(serialization) {
+  const variableMap = {};
+  serialization.variables?.forEach(variable => {
+    variableMap[variable.id] = variable.name;
+  });
+
+  /**
+   * Processes a block to replace variable IDs with variable names.
+   * Recursively checks child blocks connected to inputs or next connections.
+   * @param {object} block The block to process.
+   * @returns {object} The processed block with variable names.
+   */
+  function processBlock(block) {
```

Comment:
can we refactor this function to be outside of `getSimplifiedStateForFlyout`? That will make the code clearer.

Reply:
Yes, good idea!